enzyme/Enzyme/MLIR/Dialect/Ops.cpp

```diff
@@ -169,6 +168,272 @@ ForwardDiffOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+        auto new_dup = ActivityAttr::get(rewriter.getContext(),
+                                         Activity::enzyme_dupnoneed);
+        newRetActivityArgs.push_back(new_dup);
+      }
+
+      out_idx++;
+
+      // derivative
+      res = uop.getOutputs()[out_idx];
+      out_ty.push_back(res.getType());
+      outs_args.push_back(res);
+    };
+
+    auto epilogue = [](ForwardDiffOp &uop, ForwardDiffOp &newOp, auto &old_val,
+                       auto &new_val, auto &oldIdx, auto &newIdx) {
+      if (new_val == Activity::enzyme_dupnoneed &&
+          old_val == Activity::enzyme_dup) {
+        ++oldIdx; // skip primal
+        uop.getOutputs()[oldIdx++].replaceAllUsesWith(
+            newOp.getOutputs()[newIdx++]);
+      }
+    };
+
+    return FwdRetConverter(uop, rewriter, Activity::enzyme_dup, prologue,
+                           epilogue);
+  }
+};
+
+// Helper: check if any input is mutable.
+static inline bool isMutable(Type type) {
+  if (isa<mlir::MemRefType>(type) || isa<mlir::UnrankedMemRefType>(type) ||
+      isa<mlir::LLVM::LLVMPointerType>(type)) {
+    return true;
+  }
+
+  return false;
+}
+
+class FwdRetDupToConst final : public OpRewritePattern<ForwardDiffOp> {
+public:
+  using OpRewritePattern<ForwardDiffOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ForwardDiffOp uop,
+                                PatternRewriter &rewriter) const override {
+    auto prologue = [](ForwardDiffOp &uop, PatternRewriter &rewriter,
+                       auto &iattr, auto &res, auto &out_ty, auto &outs_args,
+                       auto &out_idx, auto &newRetActivityArgs, bool &changed) {
+      // primal
+      outs_args.push_back(res);
+      out_ty.push_back(res.getType());
+      newRetActivityArgs.push_back(iattr);
+
+      out_idx++;
+
+      // derivative
+      res = uop.getOutputs()[out_idx];
+      if (!res.use_empty()) {
+        // activity arg doesn't update
+        out_ty.push_back(res.getType());
+        outs_args.push_back(res);
+      } else {
+        // no uses, can discard
+        if (!isMutable(res.getType())) {
+          changed = true;
+          auto new_const =
+              ActivityAttr::get(rewriter.getContext(), Activity::enzyme_const);
+          newRetActivityArgs.back() = new_const;
+        }
+      }
+    };
+
+    auto epilogue = [](ForwardDiffOp &uop, ForwardDiffOp &newOp, auto &old_val,
+                       auto &new_val, auto &oldIdx, auto &newIdx) {
+      if (new_val == Activity::enzyme_const &&
+          old_val == Activity::enzyme_dup) {
+
+        uop.getOutputs()[oldIdx++].replaceAllUsesWith(
+            newOp.getOutputs()[newIdx++]);
+        ++oldIdx; // skip derivative
+      }
+    };
+
+    return FwdRetConverter(uop, rewriter, Activity::enzyme_dup, prologue,
+                           epilogue);
+  }
+};
+
+class FwdRetDupNNToConstNN final : public OpRewritePattern<ForwardDiffOp> {
+public:
+  using OpRewritePattern<ForwardDiffOp>::OpRewritePattern;
+
```

Comment:
can we simplify this by having one pattern that does all of these, and not needing to create all the separate prologue/epilogues?

Reply:
We could do that, but i.m.o. the epilogue part becomes hairy (especially if we want to handle `dup` --> `constnoneed` - you are eliminating 2 return values as compared to 1). All transforms here strictly reduce number of args by 1. 

I'll still give it a go.